File: epi_judge_python_solutions/maximum_subarray_in_circular_array.py
from test_framework import generic_test


def max_subarray_sum_in_circular(A):

    # Calculates the non-circular solution.
    def find_max_subarray():
        maximum_till = maximum = 0
        for a in A:
            maximum_till = max(a, a + maximum_till)
            maximum = max(maximum, maximum_till)
        return maximum

    # Calculates the solution which is circular.
    def find_circular_max_subarray():
        def compute_running_maximum(A):
            partial_sum = A[0]
            running_maximum = [partial_sum]
            for a in A[1:]:
                partial_sum += a
                running_maximum.append(max(running_maximum[-1], partial_sum))
            return running_maximum

        # Maximum subarray sum starts at index 0 and ends at or before index i.
        maximum_begin = compute_running_maximum(A)
        # Maximum subarray sum starts at index i + 1 and ends at the last
        # element.
        maximum_end = compute_running_maximum(A[::-1])[::-1][1:] + [0]

        # Calculates the maximum subarray which is circular.
        return max(
            begin + end for begin, end in zip(maximum_begin, maximum_end))

    return max(find_max_subarray(), find_circular_max_subarray())


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main(
            "maximum_subarray_in_circular_array.py",
            'maximum_subarray_in_circular_array.tsv',
            max_subarray_sum_in_circular))

File: epi_judge_cpp_solutions/maximum_subarray_in_circular_array.cc
#include <algorithm>
#include <vector>

#include "test_framework/generic_test.h"

using std::max;
using std::vector;

int FindMaxSubarray(const vector<int>& A);
int FindCircularMaxSubarray(const vector<int>& A);

int MaxSubarraySumInCircular(const vector<int>& A) {
  return max(FindMaxSubarray(A), FindCircularMaxSubarray(A));
}

// Calculates the non-circular solution.
int FindMaxSubarray(const vector<int>& A) {
  int maximum_till = 0, maximum = 0;
  for (int a : A) {
    maximum_till = max(a, a + maximum_till);
    maximum = max(maximum, maximum_till);
  }
  return maximum;
}

vector<int> ComputeRunningMaximum(const vector<int>& A) {
  vector<int> running_maximum;
  int sum = A.front();
  running_maximum.emplace_back(sum);
  for (int i = 1; i < size(A); ++i) {
    sum += A[i];
    running_maximum.emplace_back(max(running_maximum.back(), sum));
  }
  return running_maximum;
}

// Calculates the solution which is circular.
int FindCircularMaxSubarray(const vector<int>& A) {
  // Maximum subarray sum starts at index 0 and ends at or before index i.
  vector<int> maximum_begin = ComputeRunningMaximum(A);

  // Maximum subarray sum starts at index i + 1 and ends at the last element.
  vector<int> maximum_end = ComputeRunningMaximum({crbegin(A), crend(A)});
  maximum_end.pop_back();
  reverse(begin(maximum_end), end(maximum_end));
  maximum_end.emplace_back(0);

  // Calculates the maximum subarray which is circular.
  int circular_max = 0;
  for (int i = 0; i < size(A); ++i) {
    circular_max = max(circular_max, maximum_begin[i] + maximum_end[i]);
  }
  return circular_max;
}

// clang-format off


int main(int argc, char* argv[]) {
  std::vector<std::string> args {argv + 1, argv + argc};
  std::vector<std::string> param_names {"A"};
  return GenericTestMain(args, "maximum_subarray_in_circular_array.cc", "maximum_subarray_in_circular_array.tsv", &MaxSubarraySumInCircular,
                         DefaultComparator{}, param_names);
}
// clang-format on
